README.md
# vercoder\_inputer
Enter the verification code received by the phone or other device.  
一个基于flutter的验证码输入框控件.

## Demo  
控制用户精准输入长度.  
![][image-1]

## Features
- [x] **可以自定义验证码长度**  
- [x] **当全部输入完成后,会自动调用协议方法,通过代理返回验证码及上下文context**  
- [x] **可以通过verCode属性获取输入的验证码**
- [x] **用户自定义控件的尺寸**  


## Version
name|VercodeEditText
---|---
latest|0.8.1

## Usage
1.第一步,在你的pubspec.yml声明

	   dependencies:
	     vercoder_inputer: ^0.8.0
2.添加引用

	import 'package:vercoder_inputer/vercoder_inputer.dart';
	...
 
3.在需要使用的页面创建控件,并声明遵守协议方法  
 
```dart
class _MyHomePageState extends State < MyHomePage > implements InputerProtocol{

    //实现协议方法
    void didFinishedInputer(WGQVerCodeInputer inputer,BuildContext ctx,String verCode){
        print("verCode is $verCode");
        
        //判断验证码是否正确
        bool correct = false;//(此处修改成你自己的判断逻辑)
        if (!correct) {
        	inputer.reset();
        }
    }
 
 
 
    @override
    Widget build(BuildContext context) {
        //自定义外观样式(可选)
       Options opt = Options();
       opt.fontSize = 22.0;
       opt.fontColor = Colors.indigo;
       opt.fontWeight = FontWeight.w700;
       opt.emptyUnderLineColor = Colors.green;
       opt.inputedUnderLineColor = Colors.pink;
       opt.focusedColor = Colors.orange;
        //创建控件,并指明代理对象(delegate)
        WGQVerCodeInputer verCodeInputer = WGQVerCodeInputer(codeLength: 6, size: Size(375.0, 48.0), options:opt,delegate:this, );
        return new Scaffold(
            appBar: new AppBar(
                title: new Text(widget.title),
            ),
            body: Padding(
                padding: EdgeInsets.only(top: 100.0),
                child: verCodeInputer,
            )
        );
    }
}
```


# Contact me
- Email:  disburden@gmail.com
- blog: http://blog.wgq.name

[image-1]:	https://github.com/disburden/vercoder_inputer/blob/555db02e7ebf333cf2dc5562711eec865f184943/ScreenShots/verCode.gif?raw=true

CHANGELOG.md

# 0.8.1

* 键盘被收起后,点击输入框重新设置焦点弹出键盘

# 0.8.0

* 修正删除字符要先清除当前字符再往前移的问题.现在点击删除直接焦点退回前一个输入框
* 增加reset()方法,如果用户发现验证码不正确可以调用该方法清除所有已输入的字符.(参考美团验证码输入模式)

# 0.7.0

* 修复删除字符重新输入时,输入的位置与预期的位置不一致的问题
* Options类增加一个焦点颜色(focusedColor)的属性,可以控制当前获取焦点(正在操作)的文本框下划线的颜色.

# 0.6.1

* 修复readme文件格式错误

# 0.6.0
* 增加自定义样式的功能,通过实例化Options类来实现.

# 0.0.1
* 提交

example/test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

// import 'package:flutter/material.dart';
// import 'package:flutter_test/flutter_test.dart';

// import 'package:example/main.dart';

// void main() {
//   testWidgets('Counter increments smoke test', (WidgetTester tester) async {
//     // Build our app and trigger a frame.
//     await tester.pumpWidget(new MyApp());

//     // Verify that our counter starts at 0.
//     expect(find.text('0'), findsOneWidget);
//     expect(find.text('1'), findsNothing);

//     // Tap the '+' icon and trigger a frame.
//     await tester.tap(find.byIcon(Icons.add));
//     await tester.pump();

//     // Verify that our counter has incremented.
//     expect(find.text('0'), findsNothing);
//     expect(find.text('1'), findsOneWidget);
//   });
// }


test/vercoder_inputer_test.dart
// import 'package:test/test.dart';

// import 'package:vercoder_inputer/vercoder_inputer.dart';

// void main() {
//   test('adds one to input values', () {
// //     final calculator = new Calculator();
// //     expect(calculator.addOne(2), 3);
// //     expect(calculator.addOne(-7), -6);
// //     expect(calculator.addOne(0), 1);
// //     expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//    });
// }


lib/src/Options.dart
import 'package:flutter/material.dart';

class Options {
	double fontSize;
	FontWeight fontWeight;
	Color fontColor;
	Color emptyUnderLineColor;
	Color inputedUnderLineColor;
	Color focusedColor;

	Options({
		this.fontSize = 18.0,
		this.fontColor = Colors.black,
		this.emptyUnderLineColor = Colors.grey,
		this.inputedUnderLineColor = Colors.orange,
		this.fontWeight = FontWeight.normal,
		this.focusedColor = Colors.red,
	});

	Options merge({
		double fontSize,
		Color fontColor,
		Color emptyUnderLineColor,
		Color inputedUnderLineColor,
		FontWeight fontWeight,
		Color focusedColor,
	}) {
		return Options(
			fontSize: fontSize??this.fontSize,
			fontColor: fontColor??this.fontColor,
			emptyUnderLineColor: emptyUnderLineColor??this.emptyUnderLineColor,
			inputedUnderLineColor: inputedUnderLineColor??this.inputedUnderLineColor,
			fontWeight: fontWeight??this.fontWeight,
			focusedColor: focusedColor??this.focusedColor,
		);
	}


}

lib/src/WGQVerCodeInputer.dart
import 'package:flutter/material.dart';
import 'package:vercoder_inputer/src/Options.dart';

typedef FinishInput = void Function(WGQVerCodeInputer inputer, String verCoder, BuildContext ctx);

abstract class InputerProtocol {
	void didFinishedInputer(WGQVerCodeInputer inputer, BuildContext ctx, String verCode);
}

class WGQVerCodeInputer extends StatefulWidget {
	int codeLength = 6;
	final Size size;
	double gap = 8.0;

	// final FinishInput finishInput;

	String get verCode {
		return myState.getVerCode();
	}

	InputerProtocol delegate;

	_InputerState myState;

	String getVerCode() {
		return myState.getVerCode();
	}

	void reset(){
		myState.reset();
	}

	Options options;

	WGQVerCodeInputer({
		this.codeLength,
		this.size,
		this.options,
		this.delegate,
	}) {
		if (options == null) {
			print('need init options');
			options = Options();
		}
		this.options = options;
	}

	@override
	_InputerState createState() {
		myState = _InputerState();
		return myState;
	}
}

class _InputerState extends State <WGQVerCodeInputer> {

	List <Widget> tfs;
	List <TextEditingController> tfCtrls;
	List <String> backStrs;
	List <FocusNode> fns;

	@override
	initState() {
		super.initState();
		tfCtrls = List.generate(widget.codeLength, (index) {
			TextEditingController ctrl = TextEditingController();
			ctrl.addListener(() {});
			return ctrl;
		});

		backStrs = List.generate(widget.codeLength, (index) {
			return "●";
		});

		fns = List.generate(widget.codeLength, (index) {
			FocusNode fn = FocusNode();
			return fn;
		});

		WidgetsBinding.instance.addPostFrameCallback((_) {
			setState(() {

			});
		});
	}

	String getVerCode() {
		if (tfCtrls != null) {
			List <String> codeStr = tfCtrls.map((ctrl) {
				String char = ctrl.text == "●" ? "" : ctrl.text;
				return char;
			}).toList();
			return codeStr.join();
		}
		return "";
	}

	void reset(){
		for (TextEditingController ctrl in tfCtrls) {
			ctrl.text = "●";
		}
		FocusScope.of(context).requestFocus(fns[0]);
		setState(() {});
	}


	@override
	Widget build(BuildContext context) {
		Options opt = widget.options;
		TextField findTextFieldByIndex(int index) {
			Container tfContainer = tfs[index];
			Padding padding = tfContainer.child;
			Theme theme = padding.child;
			GestureDetector gesture = theme.child;
			Container container = gesture.child;
			TextField tf = container.child;
			return tf;
		}

		void setTextFieldSelectAll(int tfIndex) {
			TextEditingController ctrl = tfCtrls[tfIndex];
			ctrl.selection = TextSelection(baseOffset: 0, extentOffset: ctrl.text.length);
		}

		void setFocusedTextFiled(int tfIndex) {
			if (tfIndex > widget.codeLength - 1 || tfIndex < 0) {
				return;
			}
			TextField tf = findTextFieldByIndex(tfIndex);
			FocusScope.of(context).requestFocus(tf.focusNode);
			setTextFieldSelectAll(tfIndex);
		}

		void setTextFieldValue(String text, int tfIndex) {
			TextEditingController ctrl = tfCtrls[tfIndex];
			ctrl.text = text;
			setTextFieldSelectAll(tfIndex);
			setState(() {});
		}

		bool checkCode() {
			bool flag = true;
			for (TextEditingController tf in tfCtrls) {
				if (tf.text.isEmpty || tf.text == "●") {
					flag = false;
					break;
				}
			}
			return flag;
		}

		void _handleChange(String text, int index) {
			String oldtxt = backStrs[index];
			print("str:$text ${text.length} old:$oldtxt");
			if (text.length == 0) {

				if (index > 0 && oldtxt == "●") {
					setFocusedTextFiled(index - 1);
					setTextFieldValue('●', index-1);
					backStrs[index-1] = '●';
					setState(() {});
				} else {
					setTextFieldValue('●', index);
					backStrs[index] = '●';
				}
			} else {
				setTextFieldValue(text, index);
				setFocusedTextFiled(index + 1);
				backStrs[index] = "";
			}


//			String oldtxt = backStrs[index];
//			if (text.length == 0) {
//				// print('$index 文本框值发生改变,变为空');
//				setTextFieldValue('●', index);
//				// print('$index 文本框设置为默认值●');
//
//				if (index > 0 && index<widget.codeLength-1 && oldtxt == "●") {
//					// print('${index-1} 设置焦点');
//					setTextFieldValue('●', index-1);
//					setFocusedTextFiled(index - 1);
//				}
//				return;
//			} else {
//				// print('$index 文本框值发送改变,变为$text');
//				setTextFieldValue(text, index);
//				// print('${index+1} 设置焦点');
//				backStrs[index] = text;
//				setFocusedTextFiled(index + 1);
//			}

			///如果验证码全部输入完,调用回调
			if (checkCode()) {
				if (widget.delegate != null) {
					widget.delegate.didFinishedInputer(widget, context, getVerCode());
				}
			}
		}

		void focused(int index) {
			TextField tf = findTextFieldByIndex(index);
			TextEditingController ctrl = tfCtrls[index];

			if (tf.focusNode.hasFocus) {
				// print('$index 获得焦点');
				if (ctrl.text.isEmpty || ctrl.text == '●') {
					// print('$index 内容为空,设置为默认值');
					ctrl.text = '●';
				}
			}
			// print('$index 设置内容全选');
			setTextFieldSelectAll(index);
		}

		tfs = List.generate(widget.codeLength, (index) {
			FocusNode fc = fns[index];
			fc.addListener(() {
				focused(index);
			});

			///获取当前皮肤
			final ThemeData base = Theme.of(context);


			TextEditingController ctrl = tfCtrls[index];
			Color borderColor = (ctrl.text.isEmpty || ctrl.text == '●') ? opt.emptyUnderLineColor : opt.inputedUnderLineColor;
			Color textColor = (ctrl.text.isEmpty || ctrl.text == '●') ? Colors.transparent : opt.fontColor;


			if (fc.hasFocus) {
				borderColor = opt.focusedColor;
			}

			ThemeData buildDarkTheme() {
				return base.copyWith(
					textSelectionColor: Colors.transparent,
					secondaryHeaderColor: Colors.transparent,
					accentColor: Colors.white,
					inputDecorationTheme: InputDecorationTheme(
						disabledBorder: new UnderlineInputBorder(
							borderSide: new BorderSide(color: borderColor)),
					),
				);
			}

			return Container(
				width: (widget.size.width / widget.codeLength),

				child: Padding(
					padding: EdgeInsets.all(widget.gap),

					child: Theme(
						data: buildDarkTheme(),
						child: GestureDetector(
							onTap: () {
								setFocusedTextFiled(index);
							},
							behavior: HitTestBehavior.opaque,
							child: Container(
								child: TextField(
									controller: ctrl,
									key: Key(index.toString()),
									textAlign: TextAlign.center,
									keyboardType: TextInputType.number,
									maxLength: 1,
									cursorColor: Colors.transparent,
									cursorWidth: 1,
									enabled: false,
									style: TextStyle(
										fontSize: opt.fontSize,
										color: textColor,
										fontWeight: opt.fontWeight,
									),
									decoration: InputDecoration(
										counterText: "",

									),
									onChanged: (String text) {
										_handleChange(text, index);
									},
									autofocus: index == 0,
									focusNode: fns[index],

								),
							),
						),
					),
				),
			);
		});

		return SizedBox(
			height: widget.size.height,
			width: widget.size.width,
			child: ListView(
				padding: EdgeInsets.zero,
				children: tfs,
				scrollDirection: Axis.horizontal,
			),
		);
	}
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:vercoder_inputer/vercoder_inputer.dart';


void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {

	// This widget is the root of your application.
	@override
	Widget build(BuildContext context) {
		return new MaterialApp(
			title: 'Flutter Demo',
			theme: new ThemeData(
				// This is the theme of your application.
				//
				// Try running your application with "flutter run". You'll see the
				// application has a blue toolbar. Then, without quitting the app, try
				// changing the primarySwatch below to Colors.green and then invoke
				// "hot reload" (press "r" in the console where you ran "flutter run",
				// or press Run > Flutter Hot Reload in IntelliJ). Notice that the
				// counter didn't reset back to zero; the application is not restarted.
				primarySwatch: Colors.blue,
			),
			home: new MyHomePage(title: 'vercode inputer'),
		);
	}
}

class MyHomePage extends StatefulWidget {
	MyHomePage({
		Key key,
		this.title
	}): super(key: key);

	// This widget is the home page of your application. It is stateful, meaning
	// that it has a State object (defined below) that contains fields that affect
	// how it looks.

	// This class is the configuration for the state. It holds the values (in this
	// case the title) provided by the parent (in this case the App widget) and
	// used by the build method of the State. Fields in a Widget subclass are
	// always marked "final".

	final String title;

	@override
	_MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State < MyHomePage > implements InputerProtocol{
	

	void didFinishedInputer(WGQVerCodeInputer inputer,BuildContext ctx,String verCode){
		print("verCode is $verCode");
//		inputer.reset();
	}

	@override
	Widget build(BuildContext context) {
		Options opt = Options();
		opt.fontSize = 22.0;
		opt.fontColor = Colors.indigo;
		opt.fontWeight = FontWeight.w700;
		opt.emptyUnderLineColor = Colors.grey;
		opt.inputedUnderLineColor = Colors.green;
		opt.focusedColor = Colors.orange;
		WGQVerCodeInputer verCodeInputer = WGQVerCodeInputer(codeLength: 6, size: Size(375.0, 48.0), options:opt,delegate:this, );
		return new Scaffold(
			appBar: new AppBar(
				title: new Text(widget.title),
			),
			body: Padding(
				padding: EdgeInsets.only(top: 100.0),
				child: verCodeInputer,
			)
		);
	}
}